### DISCUSSION TITLE
fnames namespace

### DISCUSSION THREAD

**Discussion by vrypan** [farcasterxyz/protocol#64] [FIP Stage 1: Ideas]
I am concerned about introducing a new namespace through Farcaster.

In web2 every service introduced their own namespace and we've seen the inconveniences and problems that this caused:

- My handle on twitter is this, on instagram is that, on Skype it's an other, my email is something else.
- Every time a new service comes up there's a rush to grab the "good" handles and also a rush to grab the successful handles in other services for malicious purposes.

Also, managing a namespace comes with liabilities: People and companies will claim that they have the legal rights to a name someone else is using. How will these disputed be resolved? Under which jurisdiction? Who will judge that the use of a name was intended to mislead people? Who will resolve impersonation disputes?

Farcaster can set the paradigm for other web3 services, and I would like this paradigm to build on ENS: I own my ENS, I will be able to use my ENS as my handle in every service. On the other hand, I get some of the advantages of having a separate Farcaster Name Registry.

My core proposal is 

1. To be able to register a name on Farcaster Name Registry only if you own the corresponding .eth domain **and** it complies with the FC name restrictions (length, characters).
2. To make sure that ENS transfers are reflected to the FC registry, there can be a challenge system where anyone can challenge an fname and if it's not owned by the same address, the fname will return the zero address.
3. Successful challenges can be rewarded by the protocol by a small amount of ETH (creating an incentive keep the registry up to date).

**Comment by vrypan** [comment]
Additionally, dot (".") should be a valid character to allow ENS subdomains to be used. 

This will provide a nice way to verify that users belong to an organisation, DAO, or company.

**Comment by varunsrin** [comment]
This is a good question, and we thought long and hard about using an existing name system before deciding that our own system was the correct approach. Here's why we still believe thats the case: 
 
1. Trustworthy Names - if you join a social network, and @google and @vitalik are owned by spammers, you lose trust in the network. whats the point of a username if it doesn't really mean anything other than I was the first to get it or paid the most? ENS has this problem today. See [this account](https://app.ens.domains/address/0xA582360DA4ae1fBb912C975fcce1A0C0e2af91e8) squatting pmarca, paulg and many other names.  

2. Security - homoglyph attacks in ENS can be used to trick people into believe they are looking at a different account, and each client must attempt to implement its own solution to this. 

3. User Experience - ENS allows very long names which are difficult to handle in the context of social applications. 

I think you raise a good point that there is a "gold rush" for names if a namespace is first come first serve, and we know ENS suffers from a decent bit of squatting despite best efforts from their team to prevent it. But your proposal of 'lets use ENS' is not a fair answer to people whose usernames are being squatted already - it's only fair to people like you and me who already have our names and were good at making the gold rush!  That doesn't seem like the correct solution to the problem you raised. 

Farcaster's approach to identity is going to be two prong - a decentralized identity (the fid) and a moderated namespace (the fname). We believe that moderation of names is important and can only be done by humans for now because identifying "impersonation" and other bad actions requires cultural context and cannot be solved by mechanism design or automated rules alone. We also believe it is important for people to push new ideas for namespaces, protocols and system designs and that it is healthy and desirable for Farcaster and ENS to take different approaches to tackling what is still a very unsolved problem.

**Reply by vrypan** [reply]
2 and 3: I agree. This is why I propose that you can only register a nfname if you own the .eth AND it complies with FC rules about characters, length, etc.

1: This is the big difference.

First of all, you prioritise "old" users. The 16yr old who started their journey in web3, not web2 as a NFT artist and got known by their .eth name have lower priority compared to the 50yrs old who registered their twitter handle in the first days of twitter. Do we want this?

Then, you use examples where you think it's easy and obvious to decide. Friends, brands you are familiar with.

But it will not be always so obvious. Mr Peter Marcopoulos will request pmarca. Or a Spanish computer store called "PC Marca". Club99 in Bogota will debate Club99 with NY. I still remember how frustrating it was when the Olympics were held in Athens in 2004 and a number of local brands (from cheeses to taverns to services) could not use "Olympic" which has been their brand for decades because... it's a greek word. 

There will be a band named DonaldTrump from Ghana and an ex-POTUS. Who will decide who is the rightful owner of the handle? How? Using their personal preferences? International law? US law? Chinese? Will they ask for legal papers in order to make a decision?

And if in some cases there is a liability in making a decision, who is responsible? 

I can see why an account called "vrypan" who claims to be me but is not operated by me could be disabled. But if they claim to be "the versatile fry pan", why am I entitled to it?

What I'm saying is that we have a web3 native namespace. Let's leave it up to them to resolve these problems, this is the core reason why they are there. And at the same time, create an environment where once I register my startup's .eth domain, I can expect (ideally, be sure) that it will be respected in other services too.

**Reply by vrypan** [reply]
Also, the subdomain thing I mention, will allow users to prove they speak on behalf of or belong to organisations. This is very powerful.

**Reply by vrypan** [reply]
BTW, if FC adopts ENS and at the same time it actively blocks impersonation, it will only do good to ENS, because it reduces the value of the .eth domain who is being used for impersonation. By embracing the .eth namespace, what FC does reflects to the namespace (and I believe in a positive way).

**Reply by vrypan** [reply]
Budweiser owns beer.eth. Let's say they also register FC "beer". Then Heineken buys beer.com and make it their official site. They claim that they should also get the FC handle. Who is going to resolve this dispute? Who will receive, read and evaluate the legal papers that will be sent to (to where exactly)? Who will attend the meetings with their legal teams? And if one of them threatens to sue for liabilities, who will be responsible? The moderator who would normally deny me the use of the "google" FC handle?

**Reply by varunsrin** [reply]
I agree that a trustworthy namespace is a tough problem with a lot of subtlety and nuance and we need a good framework to solve. All the questions you have asked are valid ones. 

But the solution I'm hearing from you is "lets rely on ENS" which is a first come first serve + highest bidder system, which achieves decentralization but rewards people who were early and have money. That seems unfair in another way, why should someone who is wealthy get the name over someone else? 

ENS prioritizes decentralization over trustworthiness, and we do the opposite. There are very good reasons to pick either side of the tradeoff, but it is important to acknowledge that we cannot, by definition, have a system that is both fully decentralized and fully trustworthy.

**Reply by vrypan** [reply]
Varun, I appreciate your time to read all the above. And we agree on the tension between decentralisation and trustworthiness (though, that's what crypto solved).

If you are going with a solution that relies heavily on the human factor, then the proposed solution should not be limited to the code, but also include how this thing will be governed and operated. 

My questions above are not rhetoric, they are variations of the problems you will face and if there is no good answer this means that your proposed solution will break. How are you going to solve the problems I described? I'd really like to know the answers in order to understand your solution.

My proposal clearly describes how the handle "dwr" will be awarded: to the owner of dwr.eth. I don't understand how yours will solve the future dispute between Dan and Design Within Reach, Inc who owns dwr.com. 

Also, my proposal provides a clear process for coca-cola to register the cocacola handle on fc. Far from ideal, but very clear and well defined. How are they going to do it with yours? Like what is the process for a company or an established entity to request their handle?

Of course you could say, this is not a problem, if twitter and instagram do it, we can do it too. But these are services that you have to sign an agreement and accept their terms (which state that they can do whatever they like) in order to use them, not a permissionless protocol. 

As for user experience, the problem we are faced with is not ENS-namespace or FC-namespace. It's ENS-namespace vs ENS-namespace and FC-namespace. My initial motivation to raise the issue was that having two (actually more than two, everyone is creating their own namespace nowadays) creates additional problems. And it makes the problem of setting up a business or personal brand even harder to solve (because now, I have to find a name that is available on ENS, FC, Lenses, etc at the same time. Finding an available name becomes an n! problem with every new namespace added, and it is why so many brands don't have the same trademark, ICANN domain, twitter handle and instagram handle. And why I'm paying ICANN for my .net, .com, .org domains even though I only use one of them and why I rush to every new service to register my handle.

**Reply by vrypan** [reply]
Other than trademarks (or handles that our tech bubble considers defecto personal brands), I would also like to know how you will deal with handles that have a political or ideological context.

I can promote nazism through my casts, because it's a permissionless protocol and that's the cost of having a permissionless protocol. Can I use the Hitler handle? And if not, can I use the Stalin handle? How about binladen? TalaatPasha? How will these decisions be made? 

What about profanity? Will there be some well defined limits? Who will decide and how when the hardcore fc christians ask you to take down marywasnotavirgin in the future? Is satire allowed?

You are putting yourselves in a position nobody would like to be. You say "if I have the power, I will use it wisely", but you don't realise that a) what you think is right may not be what other people think is right and b) this power comes with a huge responsibility and costs in money, time and resources.

**Comment by varunsrin** [comment]
To wrap up our thoughts on this: 

1. Farcaster's identity system relies on a fully decentralized identifier known as the Farcaster ID. All your posts, relationships and other social information are tied to this ID which is yours for life and cannot be taken away once registered.

2. Farcaster also has an optional namespace or the Farcaster Namespace, which allows users to acquire a human readable name to use in social applications. This is not fully decentralized, and ownership is subject to a simple governance policy that will evolve over time with the goal of maximizing trustworthiness. 

Fids are the decentralized primitive that link the network, while fnames are cosmetic modifiers. This is different from ENS, DNS and many other approaches out there today where the cosmetic modifier is the link. If you acquire a name like google and this gets taken away from you for impersonation **you do not lose your social graph or content**. Everything is still tied to your fid and all your posts and relationships will continue to work with whatever new name you choose. We believe this system offers the best balance between a good user experience and decentralization. 

Another important point is that **Farcaster does not require you to use fnames**. You are free to use ENS if you prefer. You can write a client that only uses fids for identity and ens for name lookups. Our own client allows tagging by ENS today, and we encourage people to link their ENS name with their Farcaster name.

**Reply by vrypan** [reply]
> Fids are the decentralized primitive that link the network, while fnames are cosmetic modifiers

This is a very important point I have totally missed. Thank you!
